### PR TITLE
Integrate ESI into Coldfront

### LINKED ISSUE #135: Support for ESI
https://github.com/CCI-MOC/esi/issues/364

> Most things should work out of the box as ESI is OpenStack and there's support for OpenStack via the [coldfront cloud plugin](https://github.com/nerc-project/coldfront-plugin-cloud)
>
> The thing we will need to figure out is quotas for things that are specific to ESI e.g. VLAN IDs and anything else that ESI does outside of OpenStack (if any).

This is the ColdFront side of ensuring that Coldfront can support ESI. The ESI side of ensuring this communiation works can be found [here](https://github.com/CCI-MOC/esi/issues/364).

### PR DESCRIPTION
**Pull request by QuanMPhm**
Closes #135 To integrate ESI and make use of the existing code, I have made a few small restructuring choices. There’s too many to list, so I will be ready for everyone’s questions on my code changes. 

The ESI allocator object, implemented in `esi.py`, subclasses from the Openstack allocator, with a few functions overridden. The ESI allocator only makes quotas for floating IPs and networks. The appropriate Allocation Attributes have been added to `attributes.py`

The functional test for the ESI integration uses the same Openstack cluster put up by `test-py38-functional-devstack.yaml`. `ci/run_functional_tests_openstack.sh` has been edited to also create an ESI coldfront resource and to run the ESI functional tests.

### CODE REVIEW

**Review by tzumainn — COMMENTED** [review]
Hi! Just a general comment: ESI doesn't require nova/compute, so that's something you'll want to take out of the ESI quotas (and probably the "small openstack" tests).

File: src/coldfront_plugin_cloud/tests/functional/openstack_mini/test_allocation.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+from django.core.management import call_command
+from keystoneclient.v3 import client
+from cinderclient import client as cinderclient
+from neutronclient.v2_0 import client as neutronclient
+from novaclient import client as novaclient
+
+
+@unittest.skipUnless(os.getenv('FUNCTIONAL_TESTS'), 'Functional tests not enabled.')
+class TestSmallOpenstack(base.TestBase):
```
Comment (tzumainn) [inline src/coldfront_plugin_cloud/tests/functional/openstack_mini/test_allocation.py:16]: Maybe "BareMetalOpenStack"? And include identity/networking/volume?

Comment (QuanMPhm) [reply src/coldfront_plugin_cloud/tests/functional/openstack_mini/test_allocation.py:16]: That's not actually the unit test to test the ESI integration. That's just the test to make sure that Coldfront doesn't crashes if we attempt to make **Openstack** allocations on Openstack clusters that might be missing some services, like Cinder and Swift.

This is why in `ci/devstack_mini.sh`, I disabled cinder and swift.

**Comment by QuanMPhm** [comment]
@tzumainn Do you have an answer for the other questions I mentioned in my PR comments? I feel like there is a lot more information I need to know before being able to remotely continue solving this issue.

**Comment by tzumainn** [comment]
> @tzumainn Do you have an answer for the other questions I mentioned in my PR comments? I feel like there is a lot more information I need to know before being able to remotely continue solving this issue.

Whoops, sorry, I missed that there were question in the files themselves; in the future it may be easier to add questions as git comments on the PR and tag the relevant people.

Anyway, floating IPs and networks are the only items I can think of that might require allocations, since I think we explicitly do not want ColdFront to limit capacity for leasing nodes.

File: README.md [hunk 2]
```diff
@@ -111,6 +111,9 @@ usage: coldfront add_openshift_resource [-h] --name NAME --auth-url AUTH_URL [--
 coldfront add_openshift_resource: error: the following arguments are required: --name, --auth-url
 ```
 
+### Configuring for ESI
+(Quan Pham) TODO Add instructions for configuring ESI
```
Comment (QuanMPhm) [inline README.md:115]: Are there any specific instructions for me as I update the README to include information about configuring ESI?

File: ci/devstack_esi.sh [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# (Quan Pham) TODO Know we want to install in a test Devstack cluster
+# to test ESI
```
Comment (QuanMPhm) [inline ci/devstack_esi.sh:2]: What openstack services do we need for to deploy a Devstack cluster for the purpose of testing the ESI service? Do we want to test the ESI integration against more than one Devstack cluster?

File: src/coldfront_plugin_cloud/attributes.py [hunk 4]
```diff
@@ -89,6 +89,12 @@ class CloudAllocationAttribute:
 QUOTA_REQUESTS_GPU = 'OpenShift Request on GPU Quota'
 QUOTA_PVC = 'OpenShift Persistent Volume Claims Quota'
 
+###########################################################
+# ESI Quota Attributes 
+# (Quan Pham) TODO Know what ESI specific attributes would be, if any
+# From Mainn's suggesting, I've added networks and floating ips
```
Comment (QuanMPhm) [inline src/coldfront_plugin_cloud/attributes.py:95]: What ESI specific allocation attribute should be included?

File: README.md (hunk 2 above)
Comment (tzumainn) [reply README.md:115]: Ah, this I wouldn't know; I guess it would be `add_esi_resource`, since this would be distinct from OpenStack? Kristi may know for sure, though.

File: src/coldfront_plugin_cloud/esi.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+from keystoneclient.v3 import client as ks_client
+from cinderclient import client as cinderclient
+from neutronclient.v2_0 import client as neutronclient
+from novaclient import client as novaclient
+
+from coldfront_plugin_cloud import attributes, base, utils
+
+from coldfront_plugin_cloud.openstack import OpenStackResourceAllocator
+
+# (Quan Pham) TODO Know which resources ESI needs, and their multipliers
```
Comment (QuanMPhm) [inline src/coldfront_plugin_cloud/esi.py:21]: What are the multipliers and static additions for the different ESI resources? And what exactly will these resources be?

File: src/coldfront_plugin_cloud/esi.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        auth,
+        verify=os.environ.get('FUNCTIONAL_TESTS', '') != 'True'
+    )
+
+class ESIResourceAllocator(OpenStackResourceAllocator):
+
+    resource_type = 'esi'
+
+    # (Quan Pham) TODO implement setting quotas for ESI resource
+    # And know which functions needs to be implement specially for ESI
```
Comment (QuanMPhm) [inline src/coldfront_plugin_cloud/esi.py:58]: This question can be addressed after we know exactly how Coldfront should interact with an ESI cluster when an ESI allocation is approved/denied: What are all the functions that should be implement for the ESI allocator?

File: src/coldfront_plugin_cloud/management/commands/add_esi_resource.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            resource_attribute_type=ResourceAttributeType.objects.get(
+                name=attributes.RESOURCE_ROLE),
+            resource=openstack,
+            value=options['role']
+        )
+        ResourceAttribute.objects.get_or_create(
+            resource_attribute_type=ResourceAttributeType.objects.get(
+                name='quantity_label'),
+            resource=openstack,
+            value='Unit of ???' # (Quan Pham) TODO Know what each allocation unit corresponds to
```
Comment (QuanMPhm) [inline src/coldfront_plugin_cloud/management/commands/add_esi_resource.py:91]: What should be the description text for an ESI allocation unit?

File: src/coldfront_plugin_cloud/management/commands/validate_allocations.py [hunk 8]
```diff
@@ -253,3 +253,16 @@ def handle(self, *args, **options):
                             logger.warning(
                                 f"Quota for allocation {project_id} was out of date. Reapplied!"
                             )
+
+        # (Quan Pham) TODO Implement validating ESI resource
+        # Should these validation code be put in a sepearate file or 
+        # in tasks.py or utils.py as functions for each resource type?
```
Comment (QuanMPhm) [inline src/coldfront_plugin_cloud/management/commands/validate_allocations.py:259]: How exactly should we validate ESI allocations? Since the test Devstack clusters we're running to test ESI integration might not require Ironic or ESI to run on them, how should the allocations be validated without directly testing ESI functionality?

File: src/coldfront_plugin_cloud/openstack.py [hunk 9]
```diff
@@ -23,6 +23,13 @@
 
 # Map the attribute name in ColdFront, to the client of the respective
 # service, the version of the API, and the key in the payload.
+#
+# (Quan Pham) TODO: The name of the services should be the same as the
+# names or aliases listed in the Openstack Service Type Authority list 
+# (https://opendev.org/openstack/service-types-authority/)
+#
+# This is because we query the Openstack cluster to check what services 
+# are available. It is probably a bad idea to hardcode the service names here
```
Comment (QuanMPhm) [inline src/coldfront_plugin_cloud/openstack.py:32]: This is less a question, and more of a proposal for a future issue. The names of the services listed in the `QUOTA_KEY_MAPPING` dict should match what is in the Openstack Service Type Authority to allow easier tracking of what Openstack objects/services our Coldfront plugin is directly interacting with.

File: src/coldfront_plugin_cloud/tests/functional/esi/test_allocations.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+from keystoneclient.v3 import client
+from cinderclient import client as cinderclient
+from neutronclient.v2_0 import client as neutronclient
+from novaclient import client as novaclient
+
+@unittest.skipUnless(os.getenv('FUNCTIONAL_TESTS'), 'Functional tests not enabled.')
+class TestAllocation(base.TestBase):
+
+    # (Quan Pham) TODO What are the test cases for an ESI resource?
+    # What do we need to test?
```
Comment (QuanMPhm) [inline src/coldfront_plugin_cloud/tests/functional/esi/test_allocations.py:18]: How should the test cases for the ESI integration be designed?

File: ci/devstack_esi.sh (hunk 3 above)
Comment (tzumainn) [reply ci/devstack_esi.sh:2]: It depends what you're testing; ideally Keystone/Neutron/Ironic/Glance. However, if you're only testing quotas, you really only need Keystone/Neutron.

File: src/coldfront_plugin_cloud/attributes.py (hunk 4 above)
Comment (tzumainn) [reply src/coldfront_plugin_cloud/attributes.py:95]: Networks and floating IPs are all I can think of.

File: src/coldfront_plugin_cloud/esi.py (hunk 5 above)
Comment (tzumainn) [reply src/coldfront_plugin_cloud/esi.py:21]: I'm not exactly sure what multipliers and static additions mean, so this may be a question for Kristi.

**Comment by tzumainn** [comment]
I've answered the ESI specific questions; however, I'm not all that familiar with ColdFront itself, so this questions regarding allocations and the like may be better suited for Kristi.

File: src/coldfront_plugin_cloud/esi.py (hunk 5 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_cloud/esi.py:21]: Multipliers and static additions refer to what an "Allocation Unit" for an ESI allocation represents. For other resources that can be allocated like [Openstack](https://github.com/nerc-project/coldfront-plugin-cloud/blob/6228541b18ca776124d5096ff8ecca2109df1634/src/coldfront_plugin_cloud/tasks.py#L21), 1 allocation unit represented 1 CPU, 4G RAM, 20G Disk Storage, etc. If the user asked for 3 allocation units of Openstack, they would be asking for a quota of 3 CPU, 12G RAM, and so on. This means each Openstack resource (CPU, RAM, etc) has a "multiplier" (1, 4G, ...) that is used to determine how much of each resource is in one allocation unit. Additionally, each Openstack allocation has a ["static" amount of certain resources ](https://github.com/nerc-project/coldfront-plugin-cloud/blob/6228541b18ca776124d5096ff8ecca2109df1634/src/coldfront_plugin_cloud/tasks.py#L45C6-L45C15)that are always included, which are 2 Floating IPs. These are the "static additions".

This is to say, I don't know what one allocation unit of ESI would represent, and would therefore like to know the multipliers and static additions for each of its resources (floating IPs, networks, etc)

Comment (tzumainn) [reply src/coldfront_plugin_cloud/esi.py:21]: Okay, yeah, that's more of a policy question - I wouldn't be the right person to answer that.

File: README.md (hunk 2 above)
Comment (QuanMPhm) [reply README.md:115]: @knikolla Your thoughts?

File: src/coldfront_plugin_cloud/esi.py (hunk 5 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_cloud/esi.py:21]: @knikolla Based on our earlier meeting today, I'm guessing 1 network and 1 floating ip is good?

Comment (knikolla) [reply src/coldfront_plugin_cloud/esi.py:21]: Yes.

File: README.md (hunk 2 above)
Comment (knikolla) [reply README.md:115]: I guess same things that apply for OpenStack in terms of parameters, since the way Keystone is configured is the same. Just make a new command called `add_esi_resource`.

File: ci/devstack_esi.sh (hunk 3 above)
Comment (knikolla) [reply ci/devstack_esi.sh:2]: @QuanMPhm Don't even do a new Devstack cluster, just reuse the same one that is already installed in the CI and don't worry about Ironic for now.

File: src/coldfront_plugin_cloud/esi.py (hunk 6 above)
Comment (knikolla) [reply src/coldfront_plugin_cloud/esi.py:58]: Just create a new Kyestone project, users, assigns roles, same as OpenStack. Then set quotas, only neutron quotas on networks and floating ips.

File: src/coldfront_plugin_cloud/management/commands/add_esi_resource.py (hunk 7 above)
Comment (knikolla) [reply src/coldfront_plugin_cloud/management/commands/add_esi_resource.py:91]: List the quotas that are specified in the same way that we do for OpenStack. In this case it's only networks and floating ips.

File: src/coldfront_plugin_cloud/management/commands/validate_allocations.py (hunk 8 above)
Comment (knikolla) [reply src/coldfront_plugin_cloud/management/commands/validate_allocations.py:259]: Same as for OpenStack. Is project created and users added correctly. Is quota set correctly.

**Comment by QuanMPhm** [comment]
@knikolla I have finished the integration code for now. For some unrelated reason the microshift test case is not working. I will update the README once we finish code review.

File: src/coldfront_plugin_cloud/esi.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+QUOTA_KEY_MAPPING_ALL_KEYS = dict()
+for service in QUOTA_KEY_MAPPING.keys():
+    QUOTA_KEY_MAPPING_ALL_KEYS.update(QUOTA_KEY_MAPPING[service]['keys'])
+
+
+class ESIResourceAllocator(OpenStackResourceAllocator):
+
+    resource_type = 'esi'
+
+    def set_quota(self, project_id):
```
Comment (knikolla) [inline src/coldfront_plugin_cloud/esi.py:31]: I'm assuming you copied this function instead of using the one in the base class because it accesses the `QUOTA_KEY_MAPPING` from the `openstack` module.

If you made `QUOTA_KEY_MAPPING` an attribute of the `OpenStackResourceAllocator` class then you can keep overload it in this subclass and therefore keep this function in the base class without copying it.

**Review by knikolla — CHANGES_REQUESTED** [review]

File: src/coldfront_plugin_cloud/esi.py (hunk 11 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_cloud/esi.py:31]: I decided to leave `QUOTA_KEY_MAPPING` outside the class since making it a class attribute would require making small changes in many places that reference the `QUOTA_KEY_MAPPING`. Will that be fine?

File: src/coldfront_plugin_cloud/openstack.py (hunk 9 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_cloud/openstack.py]: This idea is moot for now since we no longer want to query openstack to see what services are available

File: src/coldfront_plugin_cloud/esi.py (hunk 11 above)
Comment (knikolla) [reply src/coldfront_plugin_cloud/esi.py:31]: I think those small changes are reasonable, considering they will allow us to avoid repeating the logic in `get_quota` and `set_quota`.

Comment (joachimweyl) [reply src/coldfront_plugin_cloud/esi.py:31]: @knikolla is this conversation resolved?

**Comment by joachimweyl** [comment]
@tzumainn once @knikolla approves the ColdFront side are you ready to approve the ESI side?

**Comment by tzumainn** [comment]
@joachimweyl sure - given that ESI is just OpenStack, and I don't have much expertise with ColdFront, I would assume everything just kind of works.

**Comment by joachimweyl** [comment]
@knikolla do you approve the ColdFront side?

File: src/coldfront_plugin_cloud/management/commands/add_esi_resource.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
+from django.core.management.base import BaseCommand
```
Comment (knikolla) [inline src/coldfront_plugin_cloud/management/commands/add_esi_resource.py:1]: I'd rather just use the `add_openstack_resource` function with a new flag for mentioning this is ESI.

File: src/coldfront_plugin_cloud/attributes.py [hunk 13]
```diff
@@ -89,6 +89,10 @@ class CloudAllocationAttribute:
 QUOTA_REQUESTS_GPU = 'OpenShift Request on GPU Quota'
 QUOTA_PVC = 'OpenShift Persistent Volume Claims Quota'
 
+###########################################################
+# ESI Quota Attributes 
+ESI_FLOATING_IPS = 'ESI Floating IP Quota'
```
Comment (knikolla) [inline src/coldfront_plugin_cloud/attributes.py:94]: Just use the OpenStack attribute.

File: src/coldfront_plugin_cloud/attributes.py [hunk 14]
```diff
@@ -89,6 +89,10 @@ class CloudAllocationAttribute:
...
+ESI_NETWORKS = 'ESI Network Quota'
```
Comment (knikolla) [inline src/coldfront_plugin_cloud/attributes.py:95]: Keep the attribute name as OpenStack specific rather than mentioning ESI. Don't make it automatically be added to OpenStack projects though.

File: src/coldfront_plugin_cloud/esi.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+                attributes.ESI_NETWORKS: 'network'
+            }
+        }
+    }
+
+    QUOTA_KEY_MAPPING_ALL_KEYS = {quota_key: quota_name for k in QUOTA_KEY_MAPPING.values() for quota_key, quota_name in k['keys'].items()}
+
+    resource_type = 'esi'
+
+    def get_quota(self, project_id):
```
Comment (knikolla) [inline src/coldfront_plugin_cloud/esi.py:26]: In the OpenStack base class create a new method called `get_network_quota` and reuse the same logic in both places.

**Review by knikolla — CHANGES_REQUESTED** [review]

File: src/coldfront_plugin_cloud/management/commands/add_openstack_resource.py [hunk 16]
```diff
@@ -1,3 +1,4 @@
+from sre_constants import OP_IGNORE
```
Comment (knikolla) [inline src/coldfront_plugin_cloud/management/commands/add_openstack_resource.py:1]: Is this used and if so, what is this used for?

File: src/coldfront_plugin_cloud/management/commands/validate_allocations.py [hunk 17]
```diff
@@ -66,90 +67,88 @@ def check_institution_specific_code(self, allocation, apply):
 
     def handle(self, *args, **options):
 
-        # Openstack Resources first
-        openstack_resources = Resource.objects.filter(
-            resource_type=ResourceType.objects.get(
-                name='OpenStack'
+        # Deal with Openstack and ESI resources
+        for resource_name in ['OpenStack', 'ESI']:
```
Comment (knikolla) [inline src/coldfront_plugin_cloud/management/commands/validate_allocations.py:71]: A more accurate variable name would be `resource_type`, since this corresponds to the type of resource rather than its name.

File: src/coldfront_plugin_cloud/management/commands/validate_allocations.py [hunk 18]
```diff
@@ -66,90 +67,88 @@ def check_institution_specific_code(self, allocation, apply):
...
+            openstack_resources = Resource.objects.filter(
+                resource_type=ResourceType.objects.get(
+                    name=resource_name
```
Comment (knikolla) [inline src/coldfront_plugin_cloud/management/commands/validate_allocations.py:74]: I'm curious if you're able to use something like `name__in` here. The addition of the extra level of indentation required by `for resource in [OpenStack, ESI]` messes up the diff enough to make it hard to parse what has changed.

I think you should try to split out the `for allocation in openstack_allocations` loop from inside this loop by creating an `openstack_allocations` list that has allocations for both ESI and OpenStack.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (knikolla) [reply README.md:115]: This is still unresolved. Please update or remove.

**Review by knikolla — CHANGES_REQUESTED** [review]
Almost there!

File: src/coldfront_plugin_cloud/management/commands/add_openstack_resource.py (hunk 16 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_cloud/management/commands/add_openstack_resource.py:1]: That's a mistake :P

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (knikolla) [reply README.md:118]: This is still unresolved. Please update or remove.

**Review by knikolla — APPROVED** [review]

**Review by knikolla — APPROVED** [review]

File: src/coldfront_plugin_cloud/openstack.py [hunk 19]
```diff
@@ -94,6 +62,35 @@ def get_session_for_resource(resource):
...
+        },
+        'volume': {
+            'keys': {
+                attributes.QUOTA_VOLUMES: 'volumes',
+                attributes.QUOTA_VOLUMES_GB: 'gigabytes',
+            }
+        },
+    }
+
+    QUOTA_KEY_MAPPING_ALL_KEYS = {quota_key: quota_name for k in QUOTA_KEY_MAPPING.values() for quota_key, quota_name in k['keys'].items()}
```
Comment (naved001) [inline src/coldfront_plugin_cloud/openstack.py:93]: I don't understand why this change is needed when the original code is much more readable than this.

File: src/coldfront_plugin_cloud/tests/functional/esi/test_allocations.py [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+        self.volume = cinderclient.Client(session=self.session, version=3)
+        self.networking = neutronclient.Client(session=self.session)
+        self.role_member = self.identity.roles.find(name='member')
+
+    def test_new_ESI_allocation(self):
+        user = self.new_user()
+        project = self.new_project(pi=user)
+        allocation = self.new_allocation(project, self.resource, 1)
+        allocator = esi.ESIResourceAllocator(self.resource,
+                                                         allocation)
```
Comment (naved001) [inline src/coldfront_plugin_cloud/tests/functional/esi/test_allocations.py:34]: formatting?

File: src/coldfront_plugin_cloud/openstack.py (hunk 19 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_cloud/openstack.py:93]: I wanted to turn `QUOTA_KEY_MAPPING` and `QUOTA_KEY_MAPPING_ALL_KEYS` into class attributes because I wanted the `esi.py` module to be able to override it, which would prevent some redundant code down the line. 

As for the formatting of `QUOTA_KEY_MAPPING_ALL_KEYS`, that's a formatting oversight on my part. I could make initialize it in the allocator's `__init__()` function if you'd like, though I thought that would lead to a messier diff. 

For now, I've reformatted it to look more proper.

**Comment by joachimweyl** [comment]
@naved001 is this resolved?

**Review by naved001 — APPROVED** [review]
